File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章数字三角形.py
#数字三角形  2020年第十一届省赛  lanqiaoOJ 505

n = int(input())   
a = [list(map(int, input().split())) for i in range(n)]
#数组a[][]同时当成dp[][]用
for i in range(1, n):
    for j in range(0, i + 1):
        if j == 0: a[i][j] += a[i-1][j]     # 最左边元素             
        elif j == i: a[i][j] += a[i-1][j-1] # 最右边元素            
        else:        a[i][j] += max(a[i-1][j-1:j+1])
if n & 1:  # 如果是奇数行，则返回最中间值
    print(a[-1][n // 2])
else:  # 偶数行则返回中间较大值
    print(max(a[-1][n // 2 - 1], a[-1][n // 2]))

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章数字三角形.java
//数字三角形  2020年第十一届省赛  lanqiaoOJ 505

import java.util.Scanner;
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int[][] arr = new int[N][N];
        int[][] dp = new int[N][N];
        for(int i=0;i<N;i++) {
            for(int j=0;j<=i;j++) {
                arr[i][j] = sc.nextInt();
            }
        }
        sc.close();
        dp[0][0] = arr[0][0];
        for(int i=1;i<N;i++) {
            dp[i][0] = dp[i-1][0] + arr[i][0];
        }
        for(int i=1;i<N;i++) {
            for(int j=1;j<=i;j++) {
                dp[i][j] = arr[i][j] + Math.max(dp[i-1][j], dp[i-1][j-1]);
            }
        }

        if(N%2!=0) {
            System.out.println(dp[N-1][N/2]);
        }else {
            System.out.println(Math.max(dp[N-1][N/2], dp[N-1][N/2-1]));
        }
    }
}